mmcls/core/hook/ClassNumCheckHook.py

```diff
@@ -0,0 +1,75 @@
+# Copyright (c) OpenMMLab. All rights reserved
+from operator import attrgetter
+
+from mmcv.runner import IterBasedRunner
+from mmcv.runner.hooks import HOOKS, Hook
+from mmcv.utils import is_seq_of
+
+
+@HOOKS.register_module()
+class ClassNumCheckHook(Hook):
+
+    def _check_head(self, runner, runner_dataset):
+        """Check whether the `num_classes` in head matches the length of
+        `CLASSES` in `dataset`.
+
+        Args:
+            runner (obj:`EpochBasedRunner`, `IterBasedRunner`): runner object.
+        """
+        model = runner.model
+        dataset = attrgetter(runner_dataset)(runner.data_loader)
```

Comment:
Simply passing the dataset here is more intuitive,
```suggestion
    def _check_head(self, runner, dataset):
        """Check whether the `num_classes` in head matches the length of
        `CLASSES` in `dataset`.

        Args:
            runner (obj:`EpochBasedRunner`, `IterBasedRunner`): runner object.
            dataset (obj:`BaseDataset`): the dataset to check.
        """
        model = runner.model
```
And for different runner types, just get the dataset and pass it.
```python
    def before_train_iter(self, runner):
        ...
        self._check_head(runner, runner.data_loader._dataloader.dataset)

Reply:
Yeah, I like this simplification :+1: